src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -952,20 +970,670 @@ void Chttp2ServerListener::Orphan() {
+    listener_state_ = listener_state;
+  }
+
+  const grpc_resolved_address* resolved_address() const override {
+    return &resolved_address_;
+  }
+
+  void SetOnDestroyDone(grpc_closure* on_destroy_done) override;
+
+  void Orphan() override;
+
+ private:
+  friend class experimental::PassiveListenerImpl;
+
+  class ActiveConnection : public LogicalConnection {
+   public:
+    class HandshakingState : public InternallyRefCounted<HandshakingState> {
+     public:
+      HandshakingState(RefCountedPtr<ActiveConnection> connection_ref,
+                       grpc_pollset* accepting_pollset, AcceptorPtr acceptor,
+                       const ChannelArgs& args,
+                       OrphanablePtr<grpc_endpoint> endpoint);
+
+      ~HandshakingState() override;
+
+      void Orphan() override;
+
+      void StartLocked(const ChannelArgs& args);
+
+      void ShutdownLocked(absl::Status status);
+
+      // Needed to be able to grab an external weak ref in
+      // ActiveConnection::Start()
+      using InternallyRefCounted<HandshakingState>::Ref;
+
+     private:
+      void OnTimeoutLocked();
+      static void OnReceiveSettings(void* arg, grpc_error_handle /* error */);
+      void OnHandshakeDoneLocked(absl::StatusOr<HandshakerArgs*> result);
+
+      RefCountedPtr<ActiveConnection> const connection_;
+      grpc_pollset* const accepting_pollset_;
+      const AcceptorPtr acceptor_;
+      grpc_pollset_set* const interested_parties_;
+      Timestamp const deadline_;
+      OrphanablePtr<grpc_endpoint> endpoint_;
+      // Following fields are protected by WorkSerializer.
+      RefCountedPtr<HandshakeManager> handshake_mgr_;
+      // State for enforcing handshake timeout on receiving HTTP/2 settings.
+      absl::optional<EventEngine::TaskHandle> timer_handle_;
+      grpc_closure on_receive_settings_;
+    };
+
+    ActiveConnection(RefCountedPtr<NewChttp2ServerListener> listener,
+                     grpc_pollset* accepting_pollset, AcceptorPtr acceptor,
+                     const ChannelArgs& args, MemoryOwner memory_owner,
+                     OrphanablePtr<grpc_endpoint> endpoint);
+    void Start(const ChannelArgs& args);
+
+    void SendGoAway() override;
+    void DisconnectImmediately() override;
+
+    void Orphan() override;
+
+    // Needed to be able to grab an external weak ref in
+    // NewChttp2ServerListener::OnAccept()
+    using InternallyRefCounted<LogicalConnection>::RefAsSubclass;
+
+   private:
+    static void OnClose(void* arg, grpc_error_handle error);
+    void SendGoAwayImplLocked();
+    void DisconnectImmediatelyImplLocked();
+
+    RefCountedPtr<NewChttp2ServerListener> const listener_;
+    WorkSerializer work_serializer_;
+    // Following fields are protected by WorkSerializer.
+    // Set by HandshakingState before the handshaking begins and reset when
+    // handshaking is done.
+    OrphanablePtr<HandshakingState> handshaking_state_;
```

Comment:
We can probably use `absl::variant<>` for this and `transport_`, since I think we don't use both at the same time.